rfcs/0128-redeployable-clients.md

```diff
@@ -0,0 +1,393 @@
+this information into the build/deploy steps) is that a taskcluster platform
+can evolve, with parties able to declare new services at runtime, that aren't
+necessarily able to affect the taskcluster build/deploy steps.
+
+For example, in an organisation where there is a taskcluster deployment team
+looking after a deployment, they may wish to grant scopes to another team who
+are developing additional services which plug into the core taskcluster
+deployment. By granting this other team the scopes to declare API references,
+the team can work autonomously in developing new services, without the
+taskcluster deployment team needing to restart or rebuild their taskcluster
+deployment. The second team can make their API references available and modify
+them as they see fit while the taskcluster core platform is running without any
+maintenance overhead.
+
+But we may wish to implement such a thing in a future PR - for the time being,
+the author does not care about the publish mechanism, only that the content is
+eventually served.
+
+## Changes to taskcluster client building procedure
+
+### All language clients (go, java, node.js, python)
+
+The build process for these clients should be moved into a standalone tool that
+requires a `TASKCLUSTER_ROOT_URL` to be specified to build the client for. It
+will then query the API References manifest in order to determine all the APIs
+it needs to build, and from there build packages for each API reference it
+finds.
+
+Please note, the built client will retain NO REFERENCES to the client it was
+built from, other than as comments as a reference. The value of the
+`TASKCLUSTER_ROOT_URL` at build time is simply to freeze the API references and
+schemas to a given set. A client can be built against one taskcluster
+deployment, but calling code may choose to use that client to talk to a
+different cluster. It is the responsibility of the caller to ensure that the
+methods he/she calls are compatible with the deployment he/she connects to.
+
+The API reference for a service must declare the `$schema` json schema property
+to say which API reference it implements. If this matches an API reference that
+the client builder knows how to build, it will build a source code package for
+it in a local directory. If not, it may either skip it with a warning message,
+or throw an error.
+
+This code for the code generator will live in a taskcluster repo, and only
+needs to be updated when the code generation process changes, not when service
+API definitions change.
+
+### Language clients wihtout type code generation (node.js, python)
+
+The clients that do not dynamically generate code still need to retain
+information from the API references and schemas. Those clients should be
+shipped together with a set of references and schemas, but provide the facility
+for these references and schemas to be refetched from a taskcluster deployment
+(specified by a taskcluster root URL).
```

Comment:
Do we want a client fetching lots of URLs at startup?  I think it makes more sense for the clients to be able to read a data structure or vendored file(s).